internal/gatewayapi/contexts.go

```diff
@@ -527,40 +209,100 @@ type TCPRouteContext struct {
 
 	*v1alpha2.TCPRoute
 
-	parentRefs map[v1beta1.ParentReference]*RouteParentContext
+	ParentRefs map[v1beta1.ParentReference]*RouteParentContext
+}
+
+// GetRouteType returns the Kind of the Route object, HTTPRoute,
+// TLSRoute, TCPRoute, UDPRoute etc.
+//
+// This function use the typename of RouteContext and its return is
+// corresponding to the const defined in translator, like KindHTTPRoute,
+// KindGRPCRoute, KindTLSRoute, KindTCPRoute, KindUDPRoute
+func GetRouteType(route RouteContext) v1beta1.Kind {
+	rv := reflect.ValueOf(route)
+	rt := rv.Type().String()
+	return v1beta1.Kind(strings.TrimSuffix(rt, "Context")[strings.LastIndex(rt, ".")+1:])
```

Comment:
this looks a little intimidating to me, can we use `FieldByName` here ?

Reply:
sure we can use `FieldByName`, in that case, we need to add one more field, like `type string` in struct HTTP/GRPC/.../Context, indicating the type of this struct, and all the value of field `type` comes from:

https://github.com/envoyproxy/gateway/blob/4dd4f87b8b5b83673e97ee0b77ce372cd90ce273/internal/gatewayapi/translator.go#L17-L22

@arkodg see if this is ok?

Reply:
or you could also use `GetKind()` ?
see it being used in other places in the codebase
https://github.com/envoyproxy/gateway/blob/4dd4f87b8b5b83673e97ee0b77ce372cd90ce273/internal/gatewayapi/filters.go#L769

Reply:
thanks for the heads up!

now we can use `FieldByName("Kind")`, BTW, I add some `TypeMeta` for xRoute resources in `translate` func, nor it will keep getting `""`.